pkg/remote/oci.go

```diff
@@ -25,36 +25,67 @@ import (
 	imgname "github.com/google/go-containerregistry/pkg/name"
 	v1 "github.com/google/go-containerregistry/pkg/v1"
 	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
 	"github.com/tektoncd/pipeline/pkg/client/clientset/versioned/scheme"
 )
 
-// KeychainProvider is an input to the OCIResolver which returns a keychain for fetching remote images with
-// authentication.
-type KeychainProvider func() (authn.Keychain, error)
-
 // OCIResolver will attempt to fetch Tekton resources from an OCI compliant image repository.
 type OCIResolver struct {
-	imageReference   string
-	keychainProvider KeychainProvider
+	imageReference string
+	keychain       authn.Keychain
+}
+
+func NewOCIResolver(image string, keychain authn.Keychain) OCIResolver {
+	return OCIResolver{
+		imageReference: image,
+		keychain:       keychain,
+	}
 }
 
 // GetTask will retrieve the specified task from the resolver's defined image and return its spec. If it cannot be
 // retrieved for any reason, an error is returned.
-func (o OCIResolver) GetTask(taskName string) (*v1beta1.TaskSpec, error) {
+func (o OCIResolver) GetTask(taskName string) (v1alpha1.TaskInterface, error) {
 	taskContents, err := o.readImageLayer("task", taskName)
 	if err != nil {
 		return nil, err
 	}
 
-	// Deserialize the task into a valid task spec.
-	var task v1beta1.Task
-	_, _, err = scheme.Codecs.UniversalDeserializer().Decode(taskContents, nil, &task)
+	// Deserialize the contents into a valid task spec.
+	obj, kind, err := scheme.Codecs.UniversalDeserializer().Decode(taskContents, nil, nil)
 	if err != nil {
 		return nil, fmt.Errorf("Invalid remote task %s: %w", taskName, err)
 	}
 
-	return &task.Spec, nil
+	if t, ok := obj.(*v1alpha1.Task); ok {
+		return t, err
+	}
+
+	if t, ok := obj.(*v1beta1.Task); ok {
+		return &v1alpha1.Task{
+			TypeMeta:   t.TypeMeta,
+			ObjectMeta: t.TaskMetadata(),
+			Spec: v1alpha1.TaskSpec{
+				TaskSpec: t.TaskSpec(),
+			},
+		}, nil
+	}
+
+	if t, ok := obj.(*v1alpha1.ClusterTask); ok {
+		return t, err
+	}
+
+	if t, ok := obj.(*v1beta1.ClusterTask); ok {
+		return &v1alpha1.ClusterTask{
```

Comment:
why are we converting everything to a v1alpha1 version? maybe a comment to explain

Reply:
I guess this is because those are not *passed* through the webhook. That said, we should use the `Convert*` function instead of converting it manually (mainly to re-use the code)